File: assignments/haskell/grains/grains_test.hs
import Test.HUnit (Assertion, (@=?), runTestTT, Test(..), Counts(..))
import System.Exit (ExitCode(..), exitWith)
import Grains (square, total)

exitProperly :: IO Counts -> IO ()
exitProperly m = do
  counts <- m
  exitWith $ if failures counts /= 0 || errors counts /= 0 then ExitFailure 1 else ExitSuccess

testCase :: String -> Assertion -> Test
testCase label assertion = TestLabel label (TestCase assertion)

main :: IO ()
main = exitProperly $ runTestTT $ TestList
       [ TestList grainsTests ]

i :: Integral a => a -> Integer
i = fromIntegral

grainsTests :: [Test]
grainsTests =
  [ testCase "square 1" $
    1 @=? i (square 1)
  , testCase "square 2" $
    2 @=? i (square 2)
  , testCase "square 3" $
    4 @=? i (square 3)
  , testCase "square 4" $
    8 @=? i (square 4)
  , testCase "square 16" $
    32768 @=? i (square 16)
  , testCase "square 32" $
    2147483648 @=? i (square 32)
  , testCase "square 64" $
    9223372036854775808 @=? i (square 64)
  , testCase "total grains" $
    18446744073709551615 @=? i total
  ]

File: assignments/clojure/grains/grains_test.clj
(ns grains.test (:require [clojure.test :refer :all]))
(load-file "grains.clj")

(deftest square-1
  (is (= 1 (grains/square 1))))

(deftest square-2
  (is (= 2 (grains/square 2))))

(deftest square-3
  (is (= 4 (grains/square 3))))

(deftest square-4
  (is (= 8 (grains/square 4))))

(deftest square-16
  (is (= 32768 (grains/square 16))))

(deftest square-32
  (is (= 2147483648 (grains/square 32))))

(deftest square-64
  (is (= 9223372036854775808 (grains/square 64))))

(deftest total-grains
  (is (= 18446744073709551615  (grains/total))))

(run-tests)
